tokio/src/task/scope.rs

```diff
@@ -0,0 +1,439 @@
+//! Tools for structuring concurrent tasks
+//!
+//! Tokio tasks can run completely independent of each other. However it is
+//! often useful to group tasks which try to fulfill a common goal.
+//! These groups of tasks should share the same lifetime. If the task group is
+//! no longer needed all tasks should stop. If one task errors, the other tasks
+//! might no longer be needed and should also be cancelled.
+//!
+//! The utilities inside this module allow to group tasks by following the
+//! concept of structured concurrency.
+
+use crate::{
+    sync::{CancellationToken, SharedWaitGroup},
+    task::{JoinError, JoinHandle}
+};
+use pin_project_lite::pin_project;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+#[derive(Clone)]
+struct ScopeState {
+    wait_group: SharedWaitGroup,
+    config: ScopeConfig,
+}
+
+impl ScopeState {
+    fn new(config: ScopeConfig) -> Self {
+        Self {
+            config,
+            wait_group: SharedWaitGroup::new(),
+        }
+    }
+
+    fn is_cancelled(&self) -> bool {
+        self.config.cancellation_token.is_cancelled()
+    }
+}
+
+pin_project! {
+    /// Allows to wait for a child task to join
+    pub struct ScopedJoinHandle<'scope, T> {
+        #[pin]
+        handle: JoinHandle<T>,
+        phantom: core::marker::PhantomData<&'scope ()>,
+    }
+}
+
+impl<'scope, T> Future for ScopedJoinHandle<'scope, T> {
+    // The actual type is Result<Result<T, CancellableFutureError>, JoinError>
+    // However the cancellation will only happen at the exit of the scope. This
+    // means in all cases the user still has a handle to the task, the task can
+    // not be cancelled yet.
+    type Output = Result<T, JoinError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.project().handle.poll(cx)
+    }
+}
+
+struct CancelTasksGuard<'a> {
+    scope: &'a CancellationToken,
+}
+
+impl<'a> Drop for CancelTasksGuard<'a> {
+    fn drop(&mut self) {
+        self.scope.cancel();
+    }
+}
+
+struct WaitForTasksToJoinGuard<'a> {
+    wait_group: &'a SharedWaitGroup,
+    drop_behavior: ScopeDropBehavior,
+    enabled: bool,
+}
+
+impl<'a> WaitForTasksToJoinGuard<'a> {
+    fn disarm(&mut self) {
+        self.enabled = false;
+    }
+}
+
+impl<'a> Drop for WaitForTasksToJoinGuard<'a> {
+    fn drop(&mut self) {
+        if !self.enabled {
+            return;
+        }
+
+        match self.drop_behavior {
+            ScopeDropBehavior::BlockToCompletion => {
```

Comment:
We will probably want to leave out this version in the inital implementation since it raises a few ugly qeustions - eg around what to do in a singlethreaded runtime. I just left it here so far for demo purposes.

Reply:
Agreed, and even in multi-threaded, it could be unexpected to have the threads see some hangs too.